Support multiple expandees

iTakeshi opened the pull request:
Fix `multiple expandees` error.
See #12 for the circumstances and discussions.

Notes for review:
* At first I tried to expand each `expandee` using `toMeta()`, but that attempt resulted in `java.lang.UnsupportedOperationException: Position.point on NoPosition`.
* This was because the `expandee`, which was generated in `TreeInfo.getAnnotationZippers()`, did not have an appropriate `pos: Position` (actually it was `NoPosition`), therefore I seeked a method for add the one to the `expandee`.
* I used `atPos()` method of `Positions` trait, but did it ok?
* If It is ok, please confirm followings:
  * whether I changed EVERY necessary lines in `getAnnotationZippers()` with `atPos()`
  * I used `tree.pos` as the position of annottee, but is this right?
* And if not, please give me some pointers to alternatives.

xeno-by commented:
Wow, this is great news! I'm back from the conference, and I'll do my best to review this tomorrow.

xeno-by commented on plugin/src/main/scala/org/scalamacros/paradise/typechecker/Expanders.scala:
```diff
@@ -106,24 +106,25 @@ trait Expanders {
-              case defn: scala.meta.Defn.Class => defn.copy(mods = filterMods(defn.mods))
-              case defn: scala.meta.Defn.Trait => defn.copy(mods = filterMods(defn.mods))
-              case defn: scala.meta.Defn.Object => defn.copy(mods = filterMods(defn.mods))
+            expandees.flatMap { expandee =>
```
Why use `flatMap` here? Isn't a regular `map` enough?

iTakeshi replied:
Since the returned value of this `{ expandee => ... }` block is a `List[MetaTree]`, if we use simple `map` here, `metaExpandees` will be `List[List[Tree]]` and cause a type error. We have to use `flatMap` to avoid this.

dwijnand replied:
return `metaExpandeeWithoutAnnots` instead of `List(metaExpandeeWithoutAnnots)` and `map` will work as intended.

iTakeshi replied:
Ah, that's right. We have 2 choices, which is better? (I mean, should we remove the `metaExpandeeWithoutAnnots` or leave it in the source as some information?)

```scala
// option A
val metaExpandee = toMeta(expandee)
metaExpandee.transform { ... } // this will be directly returned

// option B
val metaExpandee = toMeta(expandee)
val metaExpandeeWithoutAnnots = metaExpandee.transform { ... }
metaExpandeeWithoutAnnots
```

dwijnand replied:
Totally bike-shedding. But personally I might go for `toMeta(expandee) transform { (etc)`

xeno-by replied:
I don't mind either.

iTakeshi replied:
Ok, I'll take dwijnand's suggenstion, thanks.

xeno-by commented on tests/src/main/scala/annotations/new/main/CompanionMacros.scala:
```diff
@@ -0,0 +1,43 @@
+@compileTimeOnly("@classMacro not expanded")
+class classMacro extends scala.annotation.StaticAnnotation {
+
+  inline def apply(classDefn: Defn.Class, companionDefn: Defn.Object): Stat = meta {
```
How about we pack these two guys into one block, so that the macro works correctly both when there's a companion and when there's not?

iTakeshi replied:
Do you mean something like `inline def apply(defns: Term.Block)` ? (I got your concern, but not sure if I correctly understand your idea on implementation)

xeno-by replied:
Yes, that's right.

iTakeshi replied:
Ok, I have to leave now but I'll try that tomorrow.

iTakeshi replied:
I tried following implementation:
```scala
inline def apply(defn: Term.Block): Stat = meta {
  defn
}
```
but this failed with `wrong number of arguments` error.

The cause is that the expansion process calls `macro_annot.apply()` using `java.lang.Method.invoke` with `metaArgs: List[Tree]` (which includes `metaExpandees`) as its argument (https://github.com/scalameta/paradise/blob/2.11.8/plugin/src/main/scala/org/scalamacros/paradise/typechecker/Expanders.scala#L149). Since `metaArgs` is converted into variable-arguments by `: _*`, therefore the annotation class receives exactly 2 arguments when there is a companion object beside a class.

To avoid this, perhaps we can change the line to `newStyleMacroMeth.invoke(annotationModule, metaArgs)` and the annotation class to `class classMacro ... inline def apply(List[Stat]) ...`, but this may cause serious side-effects.

I think it is much easier to define another annotation class that takes only one argument of `Defn.Class`. 

Also, I tried defining secondary constructor like `inline def apply(classDefn: Defn.Class): Stat` to the same `classMacro` class, but this didn't compile with no stack trace available.


Are there any other possible solutions for this problem?

xeno-by replied:
I propose something simpler. What about packing expandees into a block if there are several of them?

iTakeshi replied:
Just added a commit, so please review 430af57 (hoping that I don't misunderstand your point).

Notes for review:
* I'm afraid whether this `Nil` (https://github.com/scalameta/paradise/pull/15/commits/430af577dc2aa522932c026ac8c4acd894a0452e#diff-251ef713d5e30ef24f12e79593be07e4R128) might occur some problem (although the test cases are all green).

iTakeshi commented:
Just added a commit, so please review 430af57 (hoping that I don't misunderstand your point).

Notes for review:
* I'm afraid whether this `Nil` (https://github.com/scalameta/paradise/pull/15/commits/430af577dc2aa522932c026ac8c4acd894a0452e#diff-251ef713d5e30ef24f12e79593be07e4R128) might occur some problem (although the test cases are all green).

iTakeshi commented:
Also, this implementation throws an exception when class arguments are macro-annotated like
```scala
class Qux(@argMacro id: Int) { ... }
```
This is due to that the `toMeta()` converts the argument annottee in an wrong way, and seemingly one possibility of the cause is the `pos` attached to the expandees with `atPos()` (e6b2e27) is not correct, but I couldn't get a concrete conclusion.

xeno-by commented on plugin/src/main/scala/org/scalamacros/paradise/typechecker/Expanders.scala:
```diff
@@ -106,26 +106,29 @@ trait Expanders {
           }
-          val metaArgs = metaTargs ++ metaVargss.flatten ++ metaExpandees
+          val metaArg = (metaTargs ++ metaVargss.flatten ++ metaExpandees) match {
+            case Nil => Nil
```
I think that this case is impossible. Let's remove it and let the pattern match crash instead of emitting most likely incorrect result.

xeno-by commented on plugin/src/main/scala/org/scalamacros/paradise/typechecker/Expanders.scala:
```diff
@@ -106,26 +106,29 @@ trait Expanders {
+            case Nil => Nil
+            case tree :: Nil => tree
+            case list @ _ :: tail => scala.meta.Term.Block(list.asInstanceOf[Seq[scala.meta.Stat]])
+          }
```
I suggest we only pack `metaExpandees` into a `Term.Block`. I think that type and term arguments should stay separate from the expandees. For example, `@foo(42) class C; object C` should result in `metaArgs` being `List(<42>, Term.Block(<class C>, <object C>)`. What do you think?

xeno-by commented:
@iTakeshi Sorry for the long time it took me to respond. Thanks again for your contributions and patience! I've commented inline.

Given #16, I think that we shouldn't worry about macro annotations on parameters in this pull request. 

Also, could you sign the Scala CLA at https://www.lightbend.com/contribute/cla/scala? I will need this to merge your pull request when it's ready, so we could as well parallelize reviews and signing :)

iTakeshi commented:
Thanks for reviewing,

I just signed the CLA. It is a great honor for me to have a chance to contribute scala project.

Also, I agree with removing the parameter-annotation feature (as replied to #16), so I'll leave it broken for now.

iTakeshi commented:
just added a commit.

Notes for review:
* I don't like to see `match may not be exhaustive` warning, so I changed it to throw an exception. Is it appropriate to use `MacroExpansionException` here? If not, please tell me which exception class (or object) I should use.

xeno-by commented on plugin/src/main/scala/org/scalamacros/paradise/typechecker/Expanders.scala:
```diff
@@ -124,11 +124,11 @@ trait Expanders {
-          val metaArg = (metaTargs ++ metaVargss.flatten ++ metaExpandees) match {
-            case Nil => Nil
+          val metaArgs = metaTargs ++ metaVargss.flatten ++ List(metaExpandees match {
+            case Nil => throw MacroExpansionException
```
How about `global.abort`?

xeno-by commented:
@iTakeshi Thanks a lot for your pull request! I've just merged it and released it as 3.0.0-M4.

Since you already know a lot about the plugin, would you be interested in looking into other issues? For example, #11 sounds the most pertinent.

iTakeshi commented:
Thanks for merging! I'm really happy.

OK, I'll have a try for #11 (just posted a quick question on the issue).

xeno-by commented:
@iTakeshi I am currently finishing my dissertation, and I would like to give you credit for your contributions to the implementation of new-style macros. How do I refer to you in the manuscript?

iTakeshi commented:
Finishing your PhD thesis? Wow, congratulations!! And thank you for considering to add my name to the manuscript.

* name: `Takeshi D. ITOH`
* affiliation (if necessary): `Robotic Biology Institute Inc.` (although I'm an undergrad student, this work was done as a part of my part-time job)

xeno-by commented:
Thanks! Is it okay if I write `Takeshi Itoh` to be consistent with the "first name / last name" style of mentioning other contributors in the dissertation?

iTakeshi commented:
No problem. I use `D` when I write papers just for making it easier to identify me, because there are many many `Takeshi Itoh`s in Japan :D

xeno-by commented:
Understood.
